Improve explanation for JSON string literals and processing model (#146)
oauth-wg/oauth-selective-disclosure-jwt: main.md

plain-text claim values and the salt values), and SD-JWT Releases (containing a
subset of the same mapping).

> ## The Challenge of Canonicalization

Reviewer: should this be in Implementation considerations section..?

Author: Not sure - readers might miss it there.

When receiving an SD-JWT with an associated Release, a verifier must be able to

> derive all claim values that are to be disclosed (in cleartext) and must be able
> to check that each claim value's hash was signed by the issuer.

Reviewer: I think this is the main point?
```suggestion
re-compute digests of the disclosed claim value and, given the same input values,
obtain the same digest values as signed by the issuer.
```

Usually, JSON-based formats transport claim values as simple properties of a JSON object such as this:

```
...
  "family_name": "Möbius",
  "address": {
    "street_address": "Schulstr. 12",
    "locality": "Schulpforta"
  }

> ...

Reviewer: ```suggestion
```

```

> However, a problem arises when signatures over the data need to be computed and
> verified. Common signature schemes require the same byte string as input to the

Author: ```suggestion
However, a problem arises when computation over the data need to be performed and verified, like signing or hashing. Common signature schemes require the same byte string as input to the
```

> signature verification as was used for creating the signature. In the salted

Reviewer: can this be generalized to `a problem arises when computation over the data need to be performed and verified` to cover signing and hashing from the onset?

Author: Yes, sounds good to me.

hash approach outlined above, the same problem exists: For the issuer and the
verifier to arrive at the same hash digest, the same byte string must be hashed.

JSON, however, does not prescribe a unique encoding for data, but allows for variations in the encoded string. The data above, for example, can be encoded as

```

> ...

Reviewer: ```suggestion
```

"family_name": "M\u00f6bius", 
"address": {"street_address": "Schulstr. 12", "locality": "Schulpforta"}

> ...

Author: These are intentional to show that there might be other elements here. I'm not describing a specific format, just the general approach. Note that I also didn't add commas at the end of the lines.

> ...

Reviewer: ok, let's keep this and come back if needed :)

dealt with, will lead to different hashes and the inability to verify
signatures.

There are generally two approaches to deal with this problem:

1. Canonicalization: The data is transferred in JSON format, potentially
   introducing variations in its representation, but is transformed into a
   canonical form before hashing. Both the issuer and the verifier
   must use the same canonicalization algorithm to arrive at the same byte
   string for hashing.
2. Source-string encoding: Instead of transferring data in JSON format that may
   introduce variations, the serialized data that is used as the hash input is
   transferred from the issuer to the verifier. This means that the verifier can
   easily check the hash over the byte string before deserializing the data.

Mixed approaches are conceivable, i.e., transferring both the original JSON data
plus a string suitable for hashing, but such approaches can easily lead to
undetected inconsistencies resulting in time-of-check-time-of-use type security
vulnerabilities.

In this specification, the source-string encoding approach is used, as it allows
for simple and reliable interoperability without the requirement for a

> canonicalization library. To encode the source string, JSON itself is used. To

Author: Your sentence sounds as if JSON encoding would be the only option, but there can be other approaches.

```
"family_name": "{\"s\": \"6qMQvRL5haj\", \"v\": \"M\\u00f6bius\"}"
```

Or, for the address example:
```
"address": "{\"s\": \"al1N3Zom221\", \"v\": {\"locality\": \"Schulpforta\", \"street_address\": \"Schulstr. 12\"}}"
```

A verifier can then easily check the hash over the source string before
extracting the original JSON data. Variations in the encoding of the source
string are implicitly tolerated by the verifier, as the hash is computed over a
predefined byte string and not over a JSON object.

Since the encoding is based on JSON, all value types that are allowed in JSON
are also allowed in the `v` property in the source string. This includes
numbers, strings, booleans, arrays, and objects. 

It is important to note that the SD-JWT-Release containing the source string is
neither intended nor suitable for direct consumption by an application that
needs to access the disclosed claim values. The SD-JWT-Release is only intended
to be used by a verifier to check the hash over the source string and to extract
the original JSON data. The original JSON data is then used by the application.
See (#processing_model) for details.

> Using this approach, SD-JWTs can be implemented purely based on widely available
> JSON encoding and decoding libraries. 

Reviewer: I would move this higher, where describing why this approach has been taken.
```suggestion
```

undetected inconsistencies resulting in time-of-check-time-of-use type security
vulnerabilities.

In this specification, the source string encoding approach is used, as it allows
for simple and reliable interoperability without the requirement for a
canonicalization library. To encode the source string, JSON itself is used. This
approach means that SD-JWTs can be implemented purely based on widely available
JSON encoding and decoding libraries without the need for a custom data format
for encoding data. 

To produce a source string for hashing, the data is put into a JSON object
together with the salt value, like so (non-normative example, see

> (#sd_digests_claim) for details):

Reviewer: what does `like so` mean?
typo for `like below`?

Reviewer: ah you didn't change it - will change in my next PR..

{
  "iss": "https://example.com/issuer",
  "cnf": {

>     "kty": "RSA",

Reviewer: looks like Giuseppe's PR on adding `jwk` top level claim to `cnf` has been overwritten?

Author: As above, not overwritten but not in this branch yet.

If any step fails, the input is not valid and processing MUST be aborted.

## Processing Model {#processing_model}

Neither an SD-JWT nor an SD-JWT-R is suitable for direct use by an application.
Besides the REQUIRED verification steps listed above, it is further RECOMMENDED

> that tan application-consumable format is generated from the data released in

Reviewer: ```suggestion
that an application-consumable format is generated from the data released in
```

the SD-JWT-Release. The RECOMMENDED way is to merge the released claims and any
plaintext claims in the SD-JWT recursively:

 * Objects from the released claims must be merged into existing objects from the SD-JWT. 
 * If a key is present in both objects:
   * If the value in the released claims is and object and the value in the
     SD-JWT claims is an object, the two objects MUST be merged recursively.
   * Else, the value in the released claims MUST be used.

The keys `sd_digests` and `sd_hash_alg` SHOULD be removed prior to further
processing. 

> An example can be seen in Example 3 in the Appendix.

Reviewer: why not an example with simple claim structure but complex claim structure?

Author: The complex claim structure example shows all interesting cases of the merging, especially how to merge objects. We can additionally show a simple example if you like.

}
```

## Example 3 - Complex Structured SD-JWT

In this example, a complex object such as those used for OIDC4IDA (todo reference) is used.

> In this example, the Issuer is using a following object as a set of selective disclosure claims to issue to the Holder:

Reviewer: ```suggestion
In this example, the Issuer is using a following object as a set of selectively disclosable claims to issue to the Holder:
```

}
```

## Example 2b - Mixing SD and Non-SD Claims

In this example, a variant of Example 2a, the issuer decided to apply selective
disclosure only to some of the claims. In particular, the `country` component of
the `address` is contained in the JWT as a regular claim, whereas the rest of
the claims can be disclosed selectively. Note that the processing model
described in (#processing_model) allows for merging the selectively disclosable
claims with the regular claims. 

> The JSON-payload of the JWT would look as follows:

Reviewer: ```suggestion
The JSON-payload of the SD-JWT that contains both selectively disclosable claims in the `sd_digests` object and not selectively disclosable claims in a top-level JWT claim would look as follows:
```

      "n": "pm4bOHBg-oYhAyPWzR56AWX3rUIXp11_ICDkGgS6W3ZWLts-hzwI3x65659kg4hVo9dbGoCJE3ZGF_eaetE30UhBUEgpGwrDrQiJ9zqprmcFfr3qvvkGjtth8Zgl1eM2bJcOwE7PCBHWTKWYs152R7g6Jg2OVph-a8rq-q79MhKG5QoW_mTz10QT_6H4c7PjWG1fjh8hpWNnbP_pv6d1zSwZfc5fl6yVRL0DV0V3lGHKe2Wqf_eNGjBrBLVklDTk8-stX_MWLcR-EGmXAOv0UBWitS_dXJKJu-vXJyw14nHSGuxTIK2hx1pttMft9CsvqimXKeDTU14qQL1eE7ihcw",
      "e": "AQAB"
    }
  },
  "iat": 1516239022,
  "exp": 1516247022,
  "sd_hash_alg": "sha-256",
  "sd_digests": {
    "sub": "OMdwkk2HPuiInPypWUWMxot1Y2tStGsLuIcDMjKdXMU",
    "given_name": "AfKKH4a0IZki8MFDythFaFS_Xqzn-wRvAMfiy_VjYpE",
    "family_name": "eUmXmry32JiK_76xMasagkAQQsmSVdW57Ajk18riSF0",
    "email": "-Rcr4fDyjwlM_itcMxoQZCE1QAEwyLJcibEpH114KiE",
    "phone_number": "Jv2nw0C1wP5ASutYNAxrWEnaDRIpiF0eTUAkUOp8F6Y",
    "address": {
      "street_address": "n25N6kth9N0CwjZXHeth1gfovg8_I8fGyzeY0qeLp0k",
      "locality": "gJVL_TKoT_SbA4_sv0klLTkg-YEGzVUkC-6egxegsz0",
      "region": "zXbstGPuPq2cPJfyD_-HlmqVyFMf03xH-FbeotXxdbo",
      "country": "pN-5CZ5hbumsPvLKUADm4Ott6gu0E4xj09s4Z51yb8U"
    },
    "birthdate": "UxsvgkUgPnawP6wY4hmxJ_jqiNNKni62zrX7hQOUsys"
  },
  "address": {
    "country": "US"
  }
}
```

> The holder can now, for example, release the rest of the components of the `address` claim:

Reviewer: ```suggestion
The holder can now, for example, release the rest of the components of the `address` claim in the SD-JWT-Release:
```

}
```

The issuer in this example further adds the two claims `birthdate` and `place_of_birth` to the `claims` element in plain text. The following shows the resulting SD-JWT payload:

{#example-complex-sd_jwt_payload}
```json
{
  "iss": "https://example.com/issuer",
  "cnf": {

>     "kty": "RSA",

Reviewer: why is `cnf` reverting back to not including `jwk` element..?

Author: I assume it's because this branch does not yet contain the fix for the `jwk` in `cnf`, but I'm not sure why it was in the example before. I expect that this will fix itself after merging into master.

Author: I see that the base branch was merged, that explains this: The base branch now contains the fix, this branch did not yet, so this will be fixed after merging.
